### ISSUE TITLE
Add Submit Callback to Entry

### ISSUE THREAD

**Issue by stuartmscott** [fyne-io/fyne#1683]
### Is your feature request related to a problem? Please describe:
I often encounter the UI pattern in which I'd like the user to enter a string and trigger an action. `widget.Entry` currently has the `OnChanged func(string)` callback which is triggered for every keypress, however there is not a callback for when the user has completed typing.

### Is it possible to construct a solution with the existing API?
Typically I add a `widget.Button` next to the `widget.Entry` so the user can enter a string and then tap the button to trigger the action, but this flow feels slow and clunky as the user needs to move from the keyboard to the mouse/trackpad on desktop/laptop, or from the onscreen keyboard in the bottom half of the screen to a touch-target in the top half of the screen on mobile.

### Describe the solution you'd like to see:
Add `OnSubmitted func(string)` to `widget.Entry` which, if not nil, will get called with entry.Text if;
- Entry is single line and the user presses Enter or Return
- Entry is multi-line and the user presses Shift+Enter or Shift+Return

### Mobile
Special consideration must be taken for on-screen keyboards on mobile such that the primary action button displays `Done`, `Next`, `Go` or similar. Whilst still allowing the user to enter a newline if the Entry is multi-line.

### Form
`widget.Form` could set it so `OnSubmitted` will submit the Form if the Entry is the last FormItem, otherwise it moves the focus to the next FormItem.

**Comment by Jacalz** [comment]
Just for reference, this is quite easily doable with the current API. I have it implemented as follows [Jacalz/wormhole-gui/internal/bridge/widgets/entry.go](https://github.com/Jacalz/wormhole-gui/blob/main/internal/bridge/widgets/entry.go), with an extended entry. However, I do still think that this is a great addition to the main toolkit. It is a relatively common use case that should be covered.

**Comment by adrianre12** [comment]
Just a thought, should the trigger for OnSubmitted be deselection? Pressing enter is one way to signal you have finished, but also tab or selecting another widget is also an indication that entry has finished. With that, you would need a key handler for Enter, Tab and Shift-Tab which the user can implement what gets selected next.

**Comment by stuartmscott** [comment]
Is this behaviour typical? As I write this message in GitHub, pressing Tab doesn't submit the message. Ditto for GMail and GMail classic. Similarly if I write a message in Slack, pressing Tab moves focus, pressing Enter sends it and Shift+Enter adds a newline (unfortunately :( I would prefer these last two to be switched as it is the least surprising behaviour - I can't count the number of times I've accidentally sent and unfinished message)

**Comment by adrianre12** [comment]
Ah Shift-Enter designed to make us all look unprofessional in Slack.

**Comment by adrianre12** [comment]
Bloody hell, I just pressed Shift-Enter to add a new line, grrrrrr.
From your original description, it sounded like you wanted an On Enter do something fun that you could make into a submit. I was just suggesting if you generalised it, you could also make it work for navigation, verification etc. 
I have used several applications where leaving an entry box triggers some action and also where tabbing passed the last entry takes you to the next page. OK the latter was for a high throughput data entry application but it was very useful.

**Comment by andydotxyz** [comment]
> With that, you would need a key handler for Enter, Tab and Shift-Tab which the user can implement what gets selected next.With that, you would need a key handler for Enter, Tab and Shift-Tab which the user can implement what gets selected next.

If that was what we wanted to we would use "OnFocusLost" callback instead, and the driver handles the keys for us.

**Comment by andydotxyz** [comment]
In v2.0.0 release candidate

**Comment by andydotxyz** [comment]
Still work to do on form and mobile, apologies for having close.

**Comment by andydotxyz** [comment]
Mobile keyboard considerations all implemented

**Comment by pascallohrer** [comment]
Is there a way to make this work inside `dialog.ShowForm`? As far as I can tell, the dialog uses an internal callback that I can't call from the outside?

**Comment by Jacalz** [comment]
This issue has been closed for four years. For questions, please open a GitHub discussion, ask in our Slack channel or as in a channel in our Discord server.

**Comment by pascallohrer** [comment]
Sorry, I worded my comment badly, let me try again. An earlier comment mentioned "still work to do on form [...]", and I was wondering if that part had been a) dropped because there's an out-of-the-box way to make it work in forms, b) implemented without notice (later comments only mention mobile), or c) dropped for other reasons (forgotten, not worth the effort, ...).

**Comment by andydotxyz** [comment]
You can pass an `Entry` into `ShowForm` inside a `FormItem` - it is not clear what the question is here, but it is clear that it needs a new discussion if there really is a question of usage here.

There is "more work to do in form" in the wider sense - there are many issues tracking form dialogs for next release.
